upcoming/ssh-server-in-go.md

```diff
@@ -0,0 +1,240 @@
++++
+author = ["Jiahua Chen"]
+date = "2015-12-18T04:07:33-05:00"
+title = "Writing SSH server in Go"
+series = ["Advent 2015"]
++++
+
+When I'm working on [Gogs](https://gogs.io) project, there is a need of builtin
+SSH server, which allows users to preform Git-only operations through key-based
+authentication.
+
+All available resources on the web are all minimal examples and does not fit this
+specific requirement. Therefore, I think it's worth sharing my experiences to make
+your life easier in case you just ran into same problem as mine.
+
+The code structure is pretty much same to the examples you can find on the web.
+
+1. Start a SSH listening host.
+2. Accept new requests and validate their public key with database.
+3. Preform Git operations.
+4. The most important part, return a status if no error occurs.
+
+OK, before we get started, just note that code are not supposed to be copy-paste
+and just work. It will make this post too long if involves all the details.
+
+### Prepare to start a SSH server
+
+The server must have a private key in order to start a SSH server. This is for the
+purpose of preventing [Man-in-the-middle attack](https://en.wikipedia.org/wiki/Man-in-the-middle_attack).
+
+This key does not need to be server-wide, just keep it somewhere but not in temporary
+directory because users will add this key to their `known_hosts` file.
+
+```go
+// Listen starts a SSH server listens on given port.
+func Listen(port int) {
+	config := &ssh.ServerConfig{
+		PublicKeyCallback: func(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+			pkey, err := models.SearchPublicKeyByContent(strings.TrimSpace(string(ssh.MarshalAuthorizedKey(key))))
+			if err != nil {
+                // handle error
```

Comment:
The indentation on this is spaces while the rest of the code is tabs. Can you change this to tabs. The difference causes display issues on the blog.